Sources/Bot/Services/DispatchService.swift

```diff
@@ -0,0 +1,208 @@
+import Foundation
+import Result
+import ReactiveSwift
+import ReactiveFeedback
+
+/// Orchestrates multiple merge services, one per each target branch of PRs enqueued for integration
+public final class DispatchService {
+    private let integrationLabel: PullRequest.Label
+    private let topPriorityLabels: [PullRequest.Label]
+    private let requiresAllStatusChecks: Bool
+    private let statusChecksTimeout: TimeInterval
+
+    private let logger: LoggerProtocol
+    private let gitHubAPI: GitHubAPIProtocol
+    private let scheduler: DateScheduler
+
+    /// Merge services per target branch
+    private var mergeServices: [String: MergeService]
+    private let mergeServiceLifecycleObserver: Signal<MergeServiceLifecycleEvent, NoError>.Observer?
+
+    public let healthcheck: Healthcheck
+
+    public init(
+        integrationLabel: PullRequest.Label,
+        topPriorityLabels: [PullRequest.Label],
+        requiresAllStatusChecks: Bool,
+        statusChecksTimeout: TimeInterval,
+        logger: LoggerProtocol,
+        gitHubAPI: GitHubAPIProtocol,
+        gitHubEvents: GitHubEventsServiceProtocol,
+        scheduler: DateScheduler = QueueScheduler(),
+        mergeServiceLifecycleObserver: Signal<MergeServiceLifecycleEvent, NoError>.Observer? = nil
```

Comment:
This doesn't need to be injected and defined by the caller, it's a bit strange to have it that way to be honest, `DispatchService` can create a signal and expose to be consumed by any interested parties which in this case will be `DispatchServiceContext`.

So this moves from `DispatchServiceContext` to `DispatchService` which keeps the observer private, as we don't want any other parties sending those events other than `DispatchService` itself and exposes the signal to keep the observation in `DispatchServiceContext`.

`let (lifecycleSignal, lifecycleObserver) = Signal<DispatchService.MergeServiceLifecycleEvent, NoError>.pipe()`

Reply:
Good point. That implementation comes from multiple iterations when I worked on the testing and I didn't properly pause and look at the big picture once it ended up working like I needed. Will refactor to implement as you suggested, makes more sense to me indeed.

Reply:
Fixed in a34f44573fa7b17d433abc08a60a738e8ad86306